File: 1642-furthest-building-you-can-reach/1642-furthest-building-you-can-reach.py
class Solution:
    def furthestBuilding(self, heights: List[int], bricks: int, ladders: int) -> int:
        
        import heapq as hq 
        box = [] 
        for i in range(len(heights) -1):
            gap = heights[i+1] - heights[i] 
            if gap> 0: 
                hq.heappush(box , gap)
                
                if len(box) > ladders: 
                    bricks -= hq.heappop(box) 
                    if bricks < 0: 
                        return i 
        return len(heights) -1 

File: 1642-furthest-building-you-can-reach/1642-furthest-building-you-can-reach.cpp
class Solution {
public:
    int furthestBuilding(vector<int>& height, int bricks, int ladders) {
        
        
    priority_queue<int> pq; 
    for( int i = 0;  i < height.size() -1 ; i++){ 
        int gap = height[i+1] - height[i] ;
        if( gap > 0) {
            pq.push(gap) ; 
            bricks -= gap ; 
            if( bricks < 0) { 
                ladders -- ; 
                bricks += pq.top() ; 
                pq.pop() ; 
                if ( bricks < 0 || ladders < 0) {
                    return i ; } 
            }
        }
    }
    return height.size() -1 ;

        
        
    }
};